shell/src/main/java/org/apache/kafka/shell/MetadataNodeManager.java

```diff
@@ -403,4 +429,158 @@ static List<String> clientQuotaRecordDirectories(List<EntityData> entityData) {
+     * │       │       └── ranges
+     * │       └── walMetadata
+     * ├── s3Objects
+     * │   └── 1 - object id
+     * └── s3Streams
+     *     └── 0 - stream id
+     *         ├── ranges
+     *         │   └── 0 - range index
+     *         └── s3StreamObjects
+     *             └── 3 - object id
+     * </pre>
+     */
+    private void handleExtCommitImpl(MetadataRecordType type, ApiMessage message) {
+        switch (type) {
+            // For AutoMQ-related records.
+            case S3_STREAM_RECORD: {
+                S3StreamRecord record = (S3StreamRecord) message;
+                DirectoryNode s3Streams = data.root.mkdirs("s3Streams");
+                DirectoryNode s3Stream = s3Streams.mkdirs(Long.toString(record.streamId()));
+                String data = S3StreamRecordJsonConverter.write(record, S3StreamRecord.HIGHEST_SUPPORTED_VERSION).toPrettyString();
+                s3Stream.create("data").setContents(data);
+                break;
+            }
+            case REMOVE_S3_STREAM_RECORD: {
+                RemoveS3StreamRecord record = (RemoveS3StreamRecord) message;
+                data.root.rmrf("s3Streams", Long.toString(record.streamId()));
+                break;
+            }
+            case RANGE_RECORD: {
+                RangeRecord record = (RangeRecord) message;
+                DirectoryNode streams = data.root.mkdirs("s3Streams");
+                DirectoryNode stream = streams.mkdirs(Long.toString(record.streamId()));
+                DirectoryNode ranges = stream.mkdirs("ranges");
+                FileNode file = ranges.create(Integer.toString(record.rangeIndex()));
+                file.setContents(RangeRecordJsonConverter.write(record, RangeRecord.HIGHEST_SUPPORTED_VERSION).toPrettyString());
+                break;
+            }
+            case REMOVE_RANGE_RECORD: {
+                RemoveRangeRecord record = (RemoveRangeRecord) message;
+                data.root.rmrf("s3Streams", Long.toString(record.streamId()), "ranges", Integer.toString(record.rangeIndex()));
+                break;
+            }
+            case S3_STREAM_OBJECT_RECORD: {
+                S3StreamObjectRecord record = (S3StreamObjectRecord) message;
+                DirectoryNode s3StreamObjects = data.root.mkdirs("s3Streams", Long.toString(record.streamId()), "s3StreamObjects");
+                FileNode file = s3StreamObjects.create(Long.toString(record.objectId()));
+                file.setContents(S3StreamObjectRecordJsonConverter.write(record, S3StreamObjectRecord.HIGHEST_SUPPORTED_VERSION).toPrettyString());
+                break;
+            }
+            case REMOVE_S3_STREAM_OBJECT_RECORD: {
+                RemoveS3StreamObjectRecord record = (RemoveS3StreamObjectRecord) message;
+                data.root.rmrf("s3Streams", Long.toString(record.streamId()), "s3StreamObjects", Long.toString(record.objectId()));
+                break;
+            }
+            case S3_STREAM_SET_OBJECT_RECORD: {
```

Comment:
StreamSetObjectRecord should be a file too.

Reply:
It is by design. As stated in the comment [here](https://github.com/AutoMQ/automq-for-kafka/issues/710#issuecomment-1907534321), directly saving JSON to a file results in the "ranges" content being not easily understandable by humans.